src/Poseidon/CLI/OptparseApplicativeParsers.hs

```diff
@@ -462,60 +450,59 @@ parseInGenoOne = OP.option (OP.eitherReader readGenoInput) (
+                                           (path <.> ".ind")     Nothing
+            | ext `elem` [".geno.gz", ".snp.gz"       ] =
+                Right $ GenotypeEigenstrat (path <.> ".geno.gz") Nothing
+                                           (path <.> ".snp.gz")  Nothing
+                                           (path <.> ".ind")     Nothing
+            | ext `elem` [".bed",     ".bim",   ".fam"] =
+                Right $ GenotypePlink      (path <.> ".bed")     Nothing
+                                           (path <.> ".bim")     Nothing
+                                           (path <.> ".fam")     Nothing
+            | ext `elem` [".bed.gz",  ".bim.gz"       ] =
+                Right $ GenotypePlink      (path <.> ".bed.gz")  Nothing
+                                           (path <.> ".bim.gz")  Nothing
+                                           (path <.> ".fam")     Nothing
+            | otherwise = Left $ "unknown file extension: " ++ ext
+
+parseInGenoSep :: OP.Parser GenotypeFileSpec
+parseInGenoSep = parseEigenstrat <|> parsePlink <|> parseVCF
   where
-    readGenotypeFormat :: String -> Either String GenotypeFormatSpec
-    readGenotypeFormat s = case s of
-        "EIGENSTRAT" -> Right GenotypeFormatEigenstrat
-        "PLINK"      -> Right GenotypeFormatPlink
-        _            -> Left "must be EIGENSTRAT or PLINK"
-
-parseInGenoFile :: OP.Parser FilePath
-parseInGenoFile = OP.strOption (
-    OP.long "genoFile" <>
-    OP.metavar "FILE" <>
-    OP.help "Path to the input geno file.")
-
-parseInSnpFile :: OP.Parser FilePath
-parseInSnpFile = OP.strOption (
-    OP.long "snpFile" <>
-    OP.metavar "FILE" <>
-    OP.help "Path to the input snp file.")
-
-parseInIndFile :: OP.Parser FilePath
-parseInIndFile = OP.strOption (
-    OP.long "indFile" <>
-    OP.metavar "FILE" <>
-    OP.help "Path to the input ind file.")
+    parseEigenstrat = GenotypeEigenstrat <$> parseFileWithEndings "genoFile" [".geno", ".geno.gz"] <*> pure Nothing
+                                         <*> parseFileWithEndings "snpFile"  [".snp",  ".snp.gz"]  <*> pure Nothing
+                                         <*> parseFileWithEndings "indFile"  [".ind"]              <*> pure Nothing
+    parsePlink      = GenotypeEigenstrat <$> parseFileWithEndings "bedFile"  [".bed", ".bed.gz"]   <*> pure Nothing
+                                         <*> parseFileWithEndings "bimFile"  [".bim",  ".bim.gz"]  <*> pure Nothing
+                                         <*> parseFileWithEndings "bamFile"  [".fam"]              <*> pure Nothing
+    parseVCF        = GenotypeVCF        <$> parseFileWithEndings "vcfFile"  [".vcf", ".vcf.gz"]   <*> pure Nothing
+
+parseFileWithEndings :: String -> [String] -> OP.Parser FilePath
+parseFileWithEndings long endings = OP.option (OP.maybeReader fileEndingReader) (
+    OP.long long <>
+    OP.help ("Accepted file endings are " ++ intercalate ", " endings) <>
```

Comment:
This "universal" parser function is a good idea. Scales really well. Maybe the "Accepted file endings are ..." help text is a bit too minimalist.

Reply:
True. I've now added a bit more specific help texts. Take a look.